### PR TITLE
PIN-4488 BKE - Catalog-Process - Endpoint /eservices - Applied visibility restrictions to EServices

### PR DESCRIPTION
**Pull request by nttdata-rtorsoli**

### CODE REVIEW

File: src/main/scala/it/pagopa/interop/catalogprocess/api/impl/ProcessApiServiceImpl.scala [hunk 1]
```diff
@@ -213,7 +219,9 @@ final case class ProcessApiServiceImpl(
         agreementStates <- parseArrayParameters(agreementStates)
           .traverse(AgreementState.fromValue)
           .toFuture
-        eServices       <-
+        role            <- getUserRolesFuture(contexts)
+        visibilityRestrictions = Seq(ADMIN_ROLE, API_ROLE).contains(role)
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/catalogprocess/api/impl/ProcessApiServiceImpl.scala]: if the requester is a security operator the filter is not applied and the user could see all results, while only Admin and Api operators should see drafts.

Maybe you meant it with a negation?
```suggestion
        visibilityRestrictions = !Seq(ADMIN_ROLE, API_ROLE).contains(role)
```

File: src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelCatalogQueries.scala [hunk 2]
```diff
@@ -193,8 +208,22 @@ object ReadModelCatalogQueries extends ReadModelQuery {
 
     val modeFilter = mode.map(_.toString).map(Filters.eq("data.mode", _))
 
+    val visibilityRestrictionsFilter =
+      if (visibilityRestrictions)
+        Some(
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelCatalogQueries.scala:219]: awesome! 🚀 

I think it requires also a filter based on the role.
For example using something like the `visibilityRestrictions` flag (calculated from the roles) in the query to allow ADMIN and API producers to see the eservice, but not other roles (SECURITY, SUPPORT, ...), even if they belong to the producer.

May I also ask to add a comment to explain the logic?
Something like
```
Allow only the producer with right permissions to see eservices in draft, that is:
- eservices without descriptors
- eservices with only one descriptor that is in Draft state
```

File: src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelCatalogQueries.scala [hunk 3]
```diff
@@ -193,8 +208,22 @@ object ReadModelCatalogQueries extends ReadModelQuery {
...
+          Filters.nor(
+            Filters.and(Filters.ne("data.producerId", requesterId.toString), Filters.size("data.descriptors", 0)),
+            Filters.and(
+              Filters.ne("data.producerId", requesterId.toString),
+              Filters.size("data.descriptors", 1),
+              Filters.eq("data.descriptors.state", Draft.toString())
+            )
+          )
+        )
+      else Some(Filters.empty())
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelCatalogQueries.scala]: same result, but without adding an empty condition in the query
```suggestion
      else None
```

File: src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelCatalogQueries.scala [hunk 4]
```diff
@@ -193,8 +208,22 @@ object ReadModelCatalogQueries extends ReadModelQuery {
 
     val modeFilter = mode.map(_.toString).map(Filters.eq("data.mode", _))
 
+    val visibilityRestrictionsFilter =
+      if (visibilityRestrictions)
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelCatalogQueries.scala]: this should be always applied
```suggestion
```

**Review by beetlecrunch — APPROVED** [review]

File: src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelCatalogQueries.scala [hunk 5]
```diff
@@ -193,8 +200,28 @@ object ReadModelCatalogQueries extends ReadModelQuery {
...
     val modeFilter = mode.map(_.toString).map(Filters.eq("data.mode", _))
 
+    /**
+      * Allow only the producer with right permissions to see eservices in draft, that is:
+      * - eservices without descriptors
+      * - eservices with only one descriptor that is in Draft state
+      */
+    val isSupervisorFilter =
+      Some(
+        Filters.nor(
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelCatalogQueries.scala]: may I ask you to double check the conditions?
If I didn't miss something I think for the following sets the eservice is returned while it should not:
- Producer with Security role when the eservice has no descriptors
- Producer with Security role when the eservice has only one descriptor in draft
- Consumer with Security role when the eservice has no descriptors
- Consumer with Security role when the eservice has only one descriptor in draft

File: src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelCatalogQueries.scala [hunk 6]
```diff
@@ -193,8 +200,28 @@ object ReadModelCatalogQueries extends ReadModelQuery {
...
+          Filters.and(
+            Filters.and(Filters.ne("isSuperVisor", false), Filters.ne("data.producerId", requesterId.toString)),
```
Comment (beetlecrunch) [inline src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelCatalogQueries.scala]:
```suggestion
            Filters.and(Filters.eq("isSuperVisor", true), Filters.ne("data.producerId", requesterId.toString)),
```

File: src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelCatalogQueries.scala [hunk 7]
```diff
@@ -193,8 +200,28 @@ object ReadModelCatalogQueries extends ReadModelQuery {
...
+            Filters.size("data.descriptors", 0)
+          ),
+          Filters.and(
+            Filters.and(Filters.ne("isSuperVisor", false), Filters.ne("data.producerId", requesterId.toString)),
```
Comment (beetlecrunch) [inline src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelCatalogQueries.scala]:
```suggestion
            Filters.and(Filters.eq("isSuperVisor", true), Filters.ne("data.producerId", requesterId.toString)),
```

**Review by galales — APPROVED** [review]

**Review by beetlecrunch — APPROVED** [review]

**Pull request merged**